In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats as stats
from scipy.stats import shapiro

pd.options.display.float_format = "{:.2f}".format

In [2]:
df = pd.read_excel("3. Case 2 - One-sample t-test.xlsx")

In [3]:
# Print data head
df.head()

Product ID       Product Name  Price  Volume  Volume_in_liters  \
0     5113468        Apple Pulse   4.15     500              0.50   
1     1906188       Banana Pulse   4.75     500              0.50   
2     5709623      Banana Rumble   4.30     500              0.50   
3     2642672         Berry Ride   4.55     500              0.50   
4     1997170  Blackberry Sizzle   4.20     500              0.50   

   Price_per_EQ  
0          8.30  
1          9.50  
2          8.60  
3          9.10  
4          8.40

In [4]:
df["Price_per_EQ"].describe()

count   100.00
mean      6.43
std       1.37
min       3.02
25%       5.42
50%       6.24
75%       7.43
max       9.50
Name: Price_per_EQ, dtype: float64

In [5]:
# Variable data to array
data = df["Price_per_EQ"].values

The **scipy.stats.ttest_1samp()** function from SciPy conducts a one-sample t-test. This test is used to assess if there's a significant difference between the mean of a given sample and a known population mean. 

The main parameters of the function include:


- **a**: This parameter represents the sample data as an array-like object (list, tuple, NumPy array, etc.). It’s the set of observations for which you want to perform the one-sample t-test. The sample should be a random sample from the population you want to compare it with.

- **popmean**: This is the known or hypothesized population mean value against which the sample mean is evaluated. The t-test will assess whether there's a significant difference between the sample mean and this population mean.

- **alternative**: This parameter specifies the type of alternative hypothesis to be tested. The default value is 'two-sided,' which means the test will check for a significant difference in either direction (whether the sample mean is greater or smaller than the population mean). 

Other options include 'greater' and 'less,' which test whether the sample mean is significantly greater or smaller than the population mean. As the alternative hypothesis was defined as "The sample mean PPV is greater than the population mean (6.21)," we must set this parameter to ‘greater.’

Before running the test, set the significance level alpha to 0.05.

In [6]:
# Perform one-sample t-test (one-tailed)
stats.ttest_1samp(a=data, popmean=6.21, alternative = "greater")

TtestResult(statistic=1.5858764248796753, pvalue=0.05797822784307917, df=99)

The output contains two values: the t-statistic and the p-value.

•	**statistic (t-statistic)**: The t-statistic measures how far the sample mean is from the population mean relative to the sample’s standard error. A large absolute value of the t-statistic indicates that the sample mean is far from the population mean—suggesting that there may be a significant difference between the two.

•	**pvalue (p-value)**: The p-value measures the probability of observing a test statistic as extreme or more extreme than the one obtained, assuming that the null hypothesis (i.e., no difference between the sample mean and population mean) is true. 
A small p-value indicates that the observed difference between the sample mean and population mean is unlikely to have occurred by chance alone, and we reject the null hypothesis in favor of the alternative hypothesis.


The p-value (0.0580) indicates the likelihood of obtaining a t-statistic of 1.5859 or more under the assumption that the population mean is 6.21. Given that the p-value slightly exceeds the typical significance threshold of 0.05, it fails to offer substantial evidence to reject the null hypothesis. 
Consequently, with a significance level of 5%, there’s not enough evidence to reject the null hypothesis—leading us to conclude that the sample mean is not significantly different from the population mean of 6.21, as per the results of this one-sample t-test.
In other words, since the p-value (0.0580) is greater than the alpha (0.05), we fail to reject the null hypothesis. We conclude that the mean price per volume of the products in the sample (6.43), although slightly greater than the population mean (6.21), is not *significantly* greater than this population mean at 0.05.

Next, let's revisit the t-test, evaluating our initial alternative hypothesis ("The sample mean PPV differs from the population mean (6.21)"). We'll do this by setting the 'alternative' parameter to 'two-sided.'

In [7]:
# Re-run one-sample t-test with different alternative hypothesis (two-tailed)  
stats.ttest_1samp(a=data, popmean=6.21, alternative = "two-sided")

TtestResult(statistic=1.5858764248796753, pvalue=0.11595645568615834, df=99)

Since the p-value (0.1160) exceeds the alpha (0.05), we fail to reject the null hypothesis. Accordingly, the mean price per volume of the products in the sample (6.43) is not different from the population mean at the significance level of 0.05.

We may argue that the conclusions drawn from both tests are the same. Indeed, both tests indicate that the sample mean is statistically equal to the population mean.
But this is true only if the significance level is set to 0.05. If we select a significance level of 0.10, our conclusions could potentially differ.

Indeed, the two-sided test (p=0.1160) would allow us to accept the null hypothesis and conclude that the sample mean equals the population mean. But the one-sided test (p=0.0580) would provide enough evidence to claim that the sample mean exceeds the population mean. Based on these results, we could make a completely different business decision.

This highlights the critical importance of choosing an appropriate p-value in statistical testing. It's crucial to carefully consider this value before conducting your analysis and remain consistent with it post-analysis. 
Adjusting the alpha value to fit desired outcomes after viewing test results—a practice known as ‘p-hacking’—is unethical and undermines research integrity.

Furthermore, the choice between a one-sided and two-sided test should be driven by the specific research question or hypothesis being tested. The one-sided test is more specific in its claim, while the two-sided test is broader.